pageserver/src/billing_metrics.rs

```diff
@@ -0,0 +1,285 @@
+        debug!(
+            "collected current metrics for tenant: {}: state={:?} tenant_physical_size={} remote_size={}",
+            tenant_id, tenant_state, tenant_physical_size, tenant_remote_size
+        );
+
+        current_metrics.push((
+            BillingMetricsKey {
+                tenant_id,
+                timeline_id: None,
+                metric: BillingMetricKind::PhysicalSize,
+            },
+            tenant_physical_size,
+        ));
+
+        current_metrics.push((
+            BillingMetricsKey {
+                tenant_id,
+                timeline_id: None,
+                metric: BillingMetricKind::S3StorageSize,
+            },
+            tenant_remote_size,
+        ));
+
+        // TODO add SyntheticStorageSize metric
+    }
+
+    // Filter metrics
+    let mut cached_metrics_guard = cached_metrics.lock().await;
+    let mut filtered_metrics = Vec::new();
+
+    for (curr_key, curr_val) in current_metrics.iter() {
+        if let Some(val) = cached_metrics_guard.insert(curr_key.clone(), *curr_val) {
+            if val != *curr_val {
+                // metric was updated, send it
+                filtered_metrics.push({
+                    BillingMetric::new_absolute(
+                        curr_key.metric,
+                        curr_key.tenant_id,
+                        curr_key.timeline_id,
+                        *curr_val,
+                    )
+                });
+            }
+        } else {
+            // cache the metric
+            cached_metrics_guard.insert(curr_key.clone(), *curr_val);
+            filtered_metrics.push({
+                BillingMetric::new_absolute(
+                    curr_key.metric,
+                    curr_key.tenant_id,
+                    curr_key.timeline_id,
+                    *curr_val,
+                )
+            });
+        }
+    }
+
+    if filtered_metrics.is_empty() {
+        trace!("no new metrics to send");
+        return Ok(());
+    }
+
+    // Send metrics to billing service.
+    // split into chunks of 1000 metrics to avoid exceeding the max request size
+    const CHUNK_SIZE: usize = 1000;
+    let chunks = filtered_metrics.chunks(CHUNK_SIZE);
+
+    for chunk in chunks {
+        let chunk_json = serde_json::value::to_raw_value(&EventChunk { events: chunk })
+            .expect("BillingMetric should not fail serialization");
+
+        let res = client
+            .post(metric_collection_endpoint)
+            .json(&chunk_json)
+            .send()
+            .await;
+
+        match res {
+            Ok(res) => {
+                if res.status().is_success() {
+                    debug!("metrics sent successfully, response: {:?}", res);
+                } else {
+                    error!("failed to send metrics: {:?}", res);
+                }
+            }
+            Err(err) => {
+                error!("failed to send metrics: {:?}", err);
+            }
+        }
```

Comment:
Should the chunk be attempted to be resent?

Reply:
I don't think it is necessary, because we will send metrics in other tick anyway.
Otherwise, we'll end up with too many futures spawned.

Maybe we can retry 1-2 times. WDYT?

Reply:
I guess this doesn't really matter, since you already implemented the caching for uploaded values only (comment above). I don't have an idea for this, I was just commenting to make sure it has been considered.

Reply:
I guess the retry interval actually affects this. Because if it's 1min, then retrying doesn't seem so tempting, because we will try often enough. But if it's 1h, then I think we start to want retrying.